Brokerages/Bitfinex/BaseMessage.cs

```diff
@@ -0,0 +1,137 @@
+﻿/*
+ * QUANTCONNECT.COM - Democratizing Finance, Empowering Individuals.
+ * Lean Algorithmic Trading Engine v2.0. Copyright 2014 QuantConnect Corporation.
+ * 
+ * Licensed under the Apache License, Version 2.0 (the "License"); 
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+*/
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace QuantConnect.Brokerages.Bitfinex
+{
+
+
+    /// <summary>
+    /// Base class for Bitfinex wss messages
+    /// </summary>
+    public abstract class BaseMessage
+    {
+
+        /// <summary>
+        /// Stored Keys for message
+        /// </summary>
+        protected string[] allKeys { get; set; }
+
+        /// <summary>
+        /// Stores values of message
+        /// </summary>
+        protected string[] allValues { get; set; }
+
+        /// <summary>
+        /// Creates base message instance
+        /// </summary>
+        /// <param name="values"></param>
+        public BaseMessage(string[] values)
+        {
+            allValues = values;
+        }
+
+        /// <summary>
+        /// Returns typed value from untyped json array
+        /// </summary>
+        /// <param name="key"></param>
+        /// <returns></returns>
+        public string GetString(string key)
+        {
+            return allValues[Array.IndexOf(allKeys, key)];
+        }
+
+        /// <summary>
+        /// Returns typed value from untyped json array
+        /// </summary>
+        /// <param name="key"></param>
+        /// <returns></returns>
+        public decimal GetDecimal(string key)
+        {
+            return decimal.Parse(allValues[Array.IndexOf(allKeys, key)]);
```

Comment:
It looks like the keys are always constant indexes, so we could just define constants that represent each of the keys. For example:
```C#
public class TickerMessage
{
    private static class Key
    {
        public const int ChannelID = 0;
    }
}
```
Or, I would even prefer that each message parses the keys internally so that the knowledge of keys is unknown to the outside world. It kind of appears like this is what's happening, but it's still going through this indexing system. So it looks like the indexing system is only used internally by sub-classes in order to parse the input value array. I would prefer us to document the structure of the message and then use direct array look-ups since in the scope of the `TickerMessage` constructor it's expected to know use nitty gritty message contract details while parsing the message. But if you prefer to not use the indexes directly, we can instead define the constants.

My main issue with the indexing system is that is seems like an unnecessary layer of complexity and indirection in what is otherwise a very simple type hierarchy.